Add TargetingPack and enable net46 builds for source-build.

crummel opened the pull request:
dotnet/sdk enabled net46 builds in source-build, which requires repos that sdk depends on to also bring back net46 builds.  This change enables the net46 builds, adds the TargetingPack reference, and adds the MyGet feed to download to TargetingPack.  See also: https://github.com/dotnet/source-build/issues/592, https://github.com/dotnet/sdk/pull/2309, https://github.com/dotnet/source-build/pull/588.

crummel committed:
Add TargetingPack and enable net46 builds for source-build.

crummel commented:
@nkolev92 I'm not seeing any CI running, is this the right branch?  We're intending to get this in for 2.1.1.

nkolev92 commented:
@crummel 
It doesn't run automatically, we need to trigger it with a label :) 

Why do we need to reference Microsoft.NETFramework.ReferenceAssemblies?

crummel commented:
Adding the reference assemblies enables building the full-framework version of NuGet on Linux, which we worked around earlier by just not building any net46 bits in source-build.  SDK made a change to fix a bug that requires them to build net46 bits, so we have to re-enable these builds in anything SDK depends on as well.

crummel commented:
@dotnet-bot test Apex Tests On Windows please

nkolev92 commented on build/common.project.props:
```diff
@@ -271,4 +271,8 @@
    <ApexProjects Include="$(RepositoryRootDirectory)test\NuGet.Tests.Apex\*\*.csproj" />
   </ItemGroup>
 
+  <ItemGroup>
```
I'm not a fan of having this package in every single one of our libraries. 

Especially not as dependency. 

This is a build time dependency, so it should be marked with PrivateAssets="all". 
It should also be conditioned on the tfm as it's only relevant for the .NET framework builds. 

Ideally I'd like to avoid having a copy of the package declaration in every csproj though, but I'm not sure if there's a better way until we start using Directory.Build.props everywhere. 

//cc @NuGet/nuget-client thoughts?

mishra14 replied:
Agreed. What is this package? I cant find it on NuGet Gallery.

If this is needed as a build tool then we should add it with PdbGit i.e. [here](https://github.com/NuGet/NuGet.Client/blob/dev/build/common.props#L8) and [here](https://github.com/NuGet/NuGet.Client/blob/dev/build/common.legacy.props#L7).

crummel replied:
Updated with PrivateAssets and condition, I can try referencing the package individually in csprojs if that would be better.

crummel replied:
Sorry, didn't get this update before I pushed the change.  Looking at this approach now.

crummel replied:
Moved to `common.props`, I don't think the projects using `common.legacy.props` are built in our configuration so I did not add it there.

nkolev92 commented:
@crummel 
Unfortunately we don't have dotnet bot integration :) 

The only failing tests are known flaky ones so it's all good.

crummel committed:
Address code review feedback.

nkolev92 commented on build/common.project.props:
```diff
@@ -271,4 +271,8 @@
   </ItemGroup>
 
+  <ItemGroup>
+    <PackageReference Include="Microsoft.NETFramework.ReferenceAssemblies" Version="1.0.0-alpha-003" PrivateAssets="All" Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework'" />
```
Why not `'$(TargetFramework)' == 'net46'`? 

Also you should condition it on DotNetBuildFromSource as well.

crummel replied:
Updated for both.

nkolev92 commented:
ping @crummel

crummel commented:
Sorry, had some other source-build work going on and slipped on this, I'm taking another look.

crummel committed:
Move targeting pack reference to common.props.

crummel committed:
Fix condition for including targeting pack.

crummel commented:
I've moved the reference and rebased on the latest dev.  Tests are failing for me:
```
EXEC : The path '/home/chris/nuget_NuGet.Client/packages/XunitXml.TestLogger.2.0.0/build/_common' specified in the 'TestAdapterPath' is invalid. warning : The custom test adapter search path provided was not found, provide a valid path and try again. [/home/chris/nuget_NuGet.Client/build/build.proj]
  Could not find a test logger with URI or FriendlyName 'xunit'.
/home/chris/nuget_NuGet.Client/build/build.proj(405,5): error : NETCore CoreUnitTests tests failed! Results: /home/chris/nuget_NuGet.Client/build/TestResults/CoreUnitTests-vsts.xml
```

But the build succeeds on its own - I'm running this inside a source-build now.

crummel committed:
Go back to TargetFrameworkIdentifier - there are some projects that are net40.

crummel commented:
I reverted the `TargetFrameworkIdentifier` vs `TargetFramework` change back to checking for `.NETFramework` - there are some packages building for full-frameworks other than net46 (e.g. [NuGet.Frameworks](https://github.com/NuGet/NuGet.Client/blob/cae087c24b90a708da0417179fe04d17367f0a66/src/NuGet.Core/NuGet.Frameworks/NuGet.Frameworks.csproj#L7)) so this seems like a more correct change.

crummel commented:
I reverted the `TargetFrameworkIdentifier` vs `TargetFramework` change back to checking for `.NETFramework` - there are some packages building for full-frameworks other than net46 (e.g. [NuGet.Frameworks](https://github.com/NuGet/NuGet.Client/blob/cae087c24b90a708da0417179fe04d17367f0a66/src/NuGet.Core/NuGet.Frameworks/NuGet.Frameworks.csproj#L7)) so this seems like a more correct change.

crummel commented:
@nkolev92 My source-build worked so I think this is ready for another look when you get a chance.  Thanks!

nkolev92 commented on build/common.legacy.props:
```diff
@@ -6,10 +6,13 @@
   <ItemGroup Condition="'$(Shipping)' == 'true' AND '$(IsXPlat)' != 'true'">
     <PackageReference Include="PdbGit" Version="3.0.41" PrivateAssets="all"/>
   </ItemGroup>
+  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework' AND '$(DotNetBuildFromSource)' == 'true'">
```
Does this work really work with SDK based projects?

crummel replied:
I removed this and my build is still working, so it doesn't appear to be needed anyway.  I got confused with the net40 issue.

crummel committed:
Remove targeting pack from legacy props.

crummel commented:
@nkolev92 I think I've fixed the comment you had, would you mind taking a look when you get a chance?

nkolev92 reviewed (approved):
sorry for the delay @crummel 
LGTM

nkolev92 commented:
@crummel 
Which version do you need this for?

Currently our dev is targetting 5.0 NuGet /16.0 VS which afaik will flow into 3.0 of the SDK.

crummel commented:
Thanks for getting this in!  We'd like to service this in 2.1 and 2.2 as well - I can create PRs, which branches would those be?

nkolev92 commented:
@crummel 
We don't have a branch for 2.1.500 of the SDK or 2.2 yet. 
They'll have the same NuGet bits. 

We'll need to port this when we have a branch for that.

crummel commented:
Ok, I'll log an issue on our side to track that.  Thanks!

Edit: For reference, this is issue https://github.com/dotnet/source-build/issues/672.

nkolev92 commented:
Merged to 4.9/2.1.500/2.2

nkolev92 commented:
https://github.com/NuGet/NuGet.Client/commit/42731e5ac2273616ad7075ff4fa8bdcbee7a7ccb

crummel commented:
Thanks for porting this for us!
